posthog/batch_exports/service.py

```diff
@@ -847,3 +849,71 @@ async def afetch_batch_export_runs_in_range(
+    exclude_events: collections.abc.Iterable[str] | None = None,
+    include_events: collections.abc.Iterable[str] | None = None,
+) -> dt.datetime | None:
+    """Get the earliest start_at for a batch export backfill.
+
+    If there is no data for the given model, return None.
+    """
+    interval_seconds = int(interval_time_delta.total_seconds())
+    if model == "events":
+        exclude_events = exclude_events or []
+        include_events = include_events or []
+        query = """
+            SELECT toStartOfInterval(MIN(timestamp), INTERVAL %(interval_seconds)s SECONDS)
+            FROM events
+            WHERE team_id = %(team_id)s
+            AND (length(%(include_events)s::Array(String)) = 0 OR event IN %(include_events)s::Array(String))
+            AND (length(%(exclude_events)s::Array(String)) = 0 OR event NOT IN %(exclude_events)s::Array(String))
+        """
+        query_args = {
+            "team_id": team_id,
+            "include_events": include_events,
+            "exclude_events": exclude_events,
+            "interval_seconds": interval_seconds,
+        }
+        result = sync_execute(query, query_args)[0][0]
+        # if no data, ClickHouse returns 1970-01-01 00:00:00
+        # (we just compare the year rather than the whole object because in some cases the timestamp returned by
+        # ClickHouse has a timezone and sometimes it doesn't)
+        if result.year == 1970:
+            return None
+        return result
+    elif model == "persons":
+        # In the case of persons, we need to check 2 tables: person and person_distinct_id2
+        # It's more efficient querying both tables separately and taking the minimum timestamp, rather than trying to
+        # join them together.
+        # In some cases we might have invalid timestamps, so we use an arbitrary date in the past to filter these out.
+        query = """
+            SELECT toStartOfInterval(MIN(_timestamp), INTERVAL %(interval_seconds)s SECONDS)
+            FROM person
+            WHERE team_id = %(team_id)s
+            AND _timestamp > '2000-01-01'
+            UNION ALL
+            SELECT toStartOfInterval(MIN(_timestamp), INTERVAL %(interval_seconds)s SECONDS)
+            FROM person_distinct_id2
+            WHERE team_id = %(team_id)s
+            AND _timestamp > '2000-01-01'
+        """
+        query_args = {
+            "team_id": team_id,
+            "interval_seconds": interval_seconds,
+        }
+        results = sync_execute(query, query_args)
+        # if no data, ClickHouse returns 1970-01-01 00:00:00
+        # (we just compare the year rather than the whole object because in some cases the timestamp returned by
+        # ClickHouse has a timezone and sometimes it doesn't)
+        results = [result[0] for result in results if result[0].year != 1970]
+        if not results:
+            return None
+        return min(results)
+    else:
+        raise ValueError(f"Invalid model: {model}")
```

Comment:
thought: I think we'll need a model abstraction at some point to deal with things like this.

Reply:
yeah, it's getting to the point where it would be easy to forget to update something when introducing a new model